File: src/aabb.rs
use crate::{Ray, Vector};
use std::mem;

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct Aabb {
    minimum: Vector,
    maximum: Vector,
}

impl Aabb {
    #[must_use]
    pub fn new(minimum: Vector, maximum: Vector) -> Self {
        Self { minimum, maximum }
    }

    #[must_use]
    pub fn hit(&self, r: Ray, t_min: f64, t_max: f64) -> bool {
        for a in 0..3 {
            let inv_d = 1.0 / r.direction[a];
            let mut t0 = self.minimum[a] - r.origin[a] * inv_d;
            let mut t1 = self.maximum[a] - r.origin[a] * inv_d;
            if inv_d < 0.0 {
                mem::swap(&mut t0, &mut t1);
            }
            let t_min = t_min.max(t0);
            let t_max = t_max.min(t1);
            if t_max <= t_min {
                return false;
            }
        }
        true
    }
}

File: src/aabb.cc
#include "aabb.hh"

#include <algorithm>
#include <utility>

#include "ray.hh"

using namespace rays;

namespace {
	auto hit1d(double const minimum,
	           double const maximum,
	           double const origin,
	           double const direction,
	           double &t_min,
	           double &t_max)
		-> bool
	{
		auto const inv_dir = 1.0 / direction;
		auto const [t0, t1] = std::minmax(
			(minimum - origin) * inv_dir,
			(maximum - origin) * inv_dir
		);

		t_min = std::max(t0, t_min);
		t_max = std::min(t1, t_max);

		return t_min < t_max;
	}
}

auto Aabb::hit(Ray const &r, double t_min, double t_max) const noexcept -> bool
{
	auto const [mnx, mny, mnz] = minimum_;
	auto const [mxx, mxy, mxz] = maximum_;

	auto const [ox, oy, oz] = r.origin();
	auto const [dx, dy, dz] = r.direction();

	return    hit1d(mnx, mxx, ox, dx, t_min, t_max)
	       && hit1d(mny, mxy, oy, dy, t_min, t_max)
	       && hit1d(mnz, mxz, oz, dz, t_min, t_max);
}
